mmdet/datasets/pipelines/transforms.py

```diff
@@ -494,6 +496,279 @@ def __repr__(self):
+
+        crop_rect = [pleft, ptop, pleft + swidth, ptop + sheight]
+        inter_rect = [max(0, pleft), max(0, ptop),
+                      min(ow, pleft + swidth), min(oh, ptop + sheight)]
+        dst_rect = [max(0, -pleft), max(0, -ptop),
+                    max(0, -pleft) + inter_rect[2] - inter_rect[0],
+                    max(0, -ptop) + inter_rect[3] - inter_rect[1]]
+
+        if crop_rect[0] == 0 and crop_rect[1] == 0 and \
+                crop_rect[2] == img.shape[0] and crop_rect[3] == img.shape[1]:
+            cropped = img
+        else:
+            cropped = np.zeros([sheight, swidth, 3], dtype=img.dtype)
+            cropped[:, :, ] = np.mean(img, axis=(0, 1))
+            cropped[dst_rect[1]:dst_rect[3], dst_rect[0]:dst_rect[2]] = \
+                img[inter_rect[1]:inter_rect[3], inter_rect[0]:inter_rect[2]]
+
+        return cropped
+
+
+@PIPELINES.register_module()
+class Mosaic(object):
+    """Mosaic augmentation.
+
+    Args:
+        size (tuple): Output size of the image.
+        jitter (float): It decides the size of the cropping window.
+        min_offset (float): Volume of the offset of the cropping widnwow.
+        load_pipeline (list): Pipeline for loading images and annotations.
+        sub_pipeline (list): Pipeline for other data augmentations.
+        letter_box(bool): Whether to maintain the aspect ratio of the
+            subimages.
+    """
+
+    def __init__(self, size=(640, 640), jitter=0.2, min_offset=0.2,
+                 load_pipeline=None, sub_pipeline=None, letter_box=False):
+        self.w, self.h = size
+        self.jitter = jitter
+        self.min_offset = min_offset
+        self.letter_box = letter_box
+
+        self.load_pipeline = AugCompose(load_pipeline)
+        self.sub_pipeline = AugCompose(sub_pipeline)
+
+        self.mosaic_crop = MosaicCrop()
+        self.resize = Resize(keep_ratio=False)
+
+    def __call__(self, results):
+        assert len(results) == 4
+        # TODO Mingtao: seg_field
+
+        w = self.w
+        h = self.h
+
+        cut_x = random.randint(int(w * self.min_offset), 
+                               int(w * (1 - self.min_offset)))
+        cut_y = random.randint(int(h * self.min_offset), 
+                               int(h * (1 - self.min_offset)))
+
+        tmp_img = np.zeros((h, w, 3), dtype=np.uint8)
+
+        out_bboxes = []
+        out_labels = []
+        out_ignores = []
+
+        for i_mosaic in range(4):
+
+            results_i = results[i_mosaic]
+            results_i = self.load_pipeline(results_i)
+            # TODO Mingtao: write in a abstract way
+            # e.g. with `bbox_fields` or `img_fileds`
+
+            assert results_i['img_fields'] == ['img']
+            img = results_i['img']
+            oh, ow, c = img.shape
+
+            dw = int(ow * self.jitter)
+            dh = int(oh * self.jitter)
+
+            pleft = random.randint(-dw, dw + 1)
+            pright = random.randint(-dw, dw + 1)
```

Comment:
I recommend moving this code to`MosaicCrop`.

Reply:
The variables like `pleft` are needed after `sub_pipeline`.

Reply:
The variables like `pleft` are needed after `sub_pipeline`.